src/datasets/formatting/formatting.py

```diff
@@ -306,43 +339,58 @@ def format_batch(self, pa_table: pa.Table) -> pa.Table:
         return self.simple_arrow_extractor().extract_batch(pa_table)
 
 
-class PythonFormatter(Formatter[dict, list, dict]):
-    def format_row(self, pa_table: pa.Table) -> dict:
+class PythonFormatter(Formatter[Mapping, list, Mapping]):
+    class LazyExample(LazyDict):
+        def decode(self, value, feature, key):
+            return (
+                (decode_nested_example(feature, value) if value is not None else None)
+                if self.formatter.features._column_requires_decoding[key]
+                else value
+            )
+
+    class LazyBatch(LazyDict):
+        def decode(self, value, feature, key):
+            return (
+                [decode_nested_example(feature, v) if value is not None else None for v in value]
+                if self.formatter.features._column_requires_decoding[key]
+                else value
+            )
+
+    lazy_row_type = LazyExample
+    lazy_batch_type = LazyBatch
```

Comment:
I'm all ears if you have any ideas on how to make this part cleaner.

Reply:
I think that's fine. Maybe put their definition outside of PythonFormatter for pickling reasons

And maybe rename the method "format" instead of "decode", since it not just decoding but also type formatting (i.e. torch/tf formatting)